irods/test/data_obj_test.py

```diff
@@ -2094,6 +2105,132 @@ def test_touch_operation_does_not_work_when_given_a_collection__525(self):
+            def __init__(self,pbar):
+                self.pbar = pbar
+                self.lock = threading.Lock()
+                self.total = 0
+
+            def update(self,n):
+                with self.lock:
+                    self.total += n
+                    self.pbar.update(self.total)
+
+        LEN = 1024**2*40
+        content = b'_'*LEN
+
+        progress_bar = progressbar.ProgressBar(maxval=len(content))
+        wrapped = wrapper(progress_bar)
+
+        register_update_instance(wrapped, wrapped.update)
+
+        wrapped.pbar.start()
+        self._run_pbars_for_parallel_io(content, [wrapped])
+        self.assertEqual(wrapped.pbar.currval, LEN)
+
+    @unittest.skipIf(progressbar is None, "progressbar is not installed")
+    def test_pbar_for_parallel_io_3(self):
+        from irods.manager.data_object_manager import (register_update_type, unregister_update_type)
+
+        class ProgressBar_wrapper:
+         def start(self,*x):return self.p.start(*x)
+         def update(self,*x):return self.p.update(*x)
+         @property
+         def currval(self): return self.p.currval
+         def __init__(self,*args,**kw):
+          self.p = progressbar.ProgressBar(*args,**kw)
+
+        def adapt_ProgressBar(pbar, total = weakref.WeakKeyDictionary(), # Any mutable parameters thus introduced to the function scope
+                                    l = threading.Lock()):               # are one-time initialized & memoized for subsequent calls.
+            total.setdefault(pbar, 0) # Also effectively a one-time initialization
+            def _update(n):
+                with l:
+                    total[pbar] += n
+                    pbar.update(total[pbar])
+            return _update
+
+        try:
+            actualClass = ProgressBar_wrapper # weakref-compatible wrapper for progressbar.ProgressBar
+            register_update_type(actualClass, adapt_ProgressBar)
+
+            LEN = 1024**2*40
+            content = b'_'*LEN
+            ProgressBar_ = actualClass(maxval=len(content))
+            ProgressBar_.start()
+            self._run_pbars_for_parallel_io(content, [ProgressBar_])
+            self.assertEqual(ProgressBar_.currval, LEN)
+
+        finally:
+            unregister_update_type(actualClass)
+
+    @unittest.skipIf(tqdm is None, "tqdm is not installed")
+    def test_pbar_for_parallel_io_2(self):
+        from irods.manager.data_object_manager import (register_update_type, unregister_update_type)
+
+        def adapt_tqdm(pbar, l = threading.Lock()):
+            def _update(n):
+                with l:
+                    pbar.update(n)
+            return _update
+
+        try:
+            register_update_type(tqdm.tqdm, adapt_tqdm)
+
+            LEN = 1024**2*40
+            content = b'_'*LEN
+            tqdm_1 = tqdm.tqdm(total=len(content))
+            tqdm_2 = tqdm.tqdm(total=len(content))
+            self._run_pbars_for_parallel_io(content, [tqdm_1,tqdm_2.update])
```

Comment:
Is this just demonstrating two ways of doing the same thing in the same call (e.g. passing in the callable object and the bound `update` method)?

Reply:
Yes, and that more than one progress bar can be updated concurrently with the same transfer.

Reply:
Okay. A "why" comment here might be good here, in that case.

Reply:
The updated name helps here. I think this is fine now.